### ISSUE TITLE
Add option to record tracebacks alongside query or cache records

### ISSUE THREAD

**Issue by nadege** [adamchainz/django-perf-rec#212]
In our quite large projects we sometimes have troubles understanding where a query or cache operation comes from. We sometimes record perf-rec on a whole `self.client.get()` operation which may include template rendering and so on. We sometimes spend a lot of times to debug this, leading sometimes in monkeypatching django to find where the incriminating query comes from :roll_eyes:.

I thought we could have that information directly in our records, for debug purpose only of course.

So I worked on a small POC to allow the user to define a pattern in the settings, when an operation matches that pattern, it stores the current traceback alongside the other info (sql, or keys and operation..) and displays it in the yaml record.

If you find it interesting and would like for me to work on it further to integrate the feature in `django-perf-rec` that'd be great :)

**Comment by adamchainz** [comment]
Hey  @nadege 

This is a very useful idea. I've had the pain myself in determining exactly where the responsible queryset is evaluated. Finding how it was constructed can be even worse.

Your PR is a great look at the problem, and really well made! Very thorough test coverage, beyond a POC!

The pattern-matching looks useful, but maybe it would be better to define a single callback function that returns whether or not the traceback should be recorded. It could work on DB and Cache operations equally, with the user needing to use `isinstance` to compare. To be honest, I was a bit surprised by the naming "pattern" - at first I thought it would be a regex but then on reading the code I discovered it is a simple substring check.

Using a callback would be the most flexible - allowing inspection of the type of operation,  matching multiple tables, or even a regex if the user desires. What are your thoughts?

Also perhaps there's something we can do to simplify the different kinds of operations into a more unified base? There's quite a bit of repeated logic in the PR as it stands now, maybe there's some refactoring you or I can do first before adding the feature to reduce this.

Thanks,

Adam

**Comment by nadege** [comment]
Hello! awesome :)

For the pattern matching, my first intent was indeed to use a regex but for the POC I just did a substring check. I will look into using a callback though, it's a great idea. 

A common base for the different operations would be nice too. I let you know if I find time to work on that later this week. But don't hesitate to do it if you find time of course :smile:

**Comment by nadege** [comment]
Hello again :) 

Now that the refactor is done, I'm looking into this again. Could you give me more detail on the callback idea you had, if you remember ?
What I understand/imagine is: a function written in the user codebase, its path passed to django-perf-rec via a settings.
The setting default value would be a function returning false, so we don't record the traceback by default.
The callback would be called inside the `on_op` function, with the operation as parameter. 

Like this:
```python
    def on_op(self, op):
        op_record = {op.name: op.query}

        callback = import_function(settings.TRACEBACK_CALLBACK)
        if callback(op):
            op_record["traceback"] = traceback.StackSummary.from_list(op.tb).format()

        self.record.append(op_record)
```

Or was it something else?

**Comment by adamchainz** [comment]
I think an argument to `record()` would be the

**Comment by nadege** [comment]
Ah yes nice idea, working on it :)

**Comment by adamchainz** [comment]
🎉 Released in version 4.9.0, from PR #299.

Thanks for your work and persistence on this @nadege , definitely a useful addition!

**Comment by nadege** [comment]
Awesome! I'm glad this was merged :)
Thank you for giving details about your changes. 
And I learned about `mock.patch.object` :tada:
